### PR TITLE
Apply transitions to snapshotted contents instead of showing invalid render targets on iOS as an orientation change is in progress.

### PR DESCRIPTION
**Pull request by chinmaygarde**
Fixes https://github.com/flutter/flutter/issues/16322.

It is extremely hard to make the transitions seamless across devices and simulators as the behavior is so different on both. I have attempted to find the most seamless solution on both. The transitions definitely look better on devices because of the content interpolation however. The black bars and other rendering artifacts (jumping between sizes and flashing) mentioned in the linked ticket are fixed on both variants however.

### CODE REVIEW

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm [hunk 1]
```diff
@@ -1197,4 +1197,87 @@ - (NSObject*)valuePublishedByPlugin:(NSString*)pluginKey {
...
+  [super viewWillTransitionToSize:size withTransitionCoordinator:coordinator];
+
+  // On transition behavior on iOS and how Flutter tries to reconcile differences.
+  // -----------------------------------------------------------------------------
+  //
+  // The point of this method is to hide rendering artifacts as the old backing store for the
+  // renderer is being torn down and recreated at the new size. Since Flutter won't generating
+  // frames in transitionary state states, snpashotted contents have to interpolated instead.
+  //
+  // iOS seems the layout the view in the final state after the transition, snapshot it, and animate
```
Comment (dnfield) [inline shell/platform/darwin/ios/framework/Source/FlutterViewController.mm]: `iOS sees`?

Comment (dnfield) [reply shell/platform/darwin/ios/framework/Source/FlutterViewController.mm]: `snapshots` `animates`?

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm [hunk 2]
```diff
@@ -1197,4 +1197,87 @@ - (NSObject*)valuePublishedByPlugin:(NSString*)pluginKey {
...
+  // the simulator, the view seems to be resized. Do what looks least jerky on both devices and
+  // simulators. Either way, the overlay has to be removed and discarded.
+#if TARGET_OS_SIMULATOR
+  dispatch_after(dispatch_time(DISPATCH_TIME_NOW,
+                               static_cast<int64_t>(coordinator.transitionDuration * NSEC_PER_SEC)),
+                 dispatch_get_main_queue(), ^{
+                   [snapshot removeFromSuperview];
+                 });
+#else   // TARGET_OS_SIMULATOR
+  [UIView animateWithDuration:0.25
```
Comment (dnfield) [inline shell/platform/darwin/ios/framework/Source/FlutterViewController.mm]: Why 0.25?

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm (hunk 1 above)
Comment (chinmaygarde) [reply shell/platform/darwin/ios/framework/Source/FlutterViewController.mm]: "iOS seems to layout the". Fixed. Also fixed the rest.

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm (hunk 2 above)
Comment (chinmaygarde) [reply shell/platform/darwin/ios/framework/Source/FlutterViewController.mm]: Dunno. Looked right. It's totally a magic number. We could make this configurable the embedder but the different behavior between devices and simulators is really hard to reconcile in a single API.

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm [hunk 3]
```diff
@@ -1197,4 +1197,87 @@ - (NSObject*)valuePublishedByPlugin:(NSString*)pluginKey {
...
+  // inflating a view when not necessary.
+  if (!self.isViewLoaded) {
+    return;
+  }
+
+  // There is no transition and hence nothing to do.
+  if (CGSizeEqualToSize(self.view.bounds.size, size)) {
+    return;
+  }
+
```
Comment (xster) [inline shell/platform/darwin/ios/framework/Source/FlutterViewController.mm]: also check that _splashScreenView is null since these 2 things can collide?

**Comment by dnfield** [comment]
Maybe we can sit together on this today at some point - I tried building it locally and I'm not seeing any noticable difference on a physical iPhone :(

**Comment by chinmaygarde** [comment]
Admittedly, I was primarily testing on the iPad on which it looks great. I tried it on the iPhone just not where there is some stretching. However, the flashing and the black bars are no longer present. Are you sure you are pushing updated builds? Here is what I see https://youtu.be/pNvAbmcOdoI.

I could also be like just host the simulators and the iPad have different behavior, the iPhone handles it differently as well.

**Comment by dnfield** [comment]
I see what you see in your video without your changes, on an iPhone 11.

**Comment by chinmaygarde** [comment]
> I see what you see in your video without your changes, on an iPhone 11.

Wait, you didn't see the black bars and jumping around like in the Gif in the linked issue? I thought I definitely saw that on the device before I attempted this. Thats what I was primarily attempting to fix.

**Comment by dnfield** [comment]
The gif is on a simulator.  I assumed it was on a physical device.  It's also possible that it was on an earlier version of iOS (or an earlier iPhone???).  Or I just have something confused...

**Comment by adriancmurray** [comment]
> ...the iPhone handles it differently as well.

The iPad rotation looks fantastic. I am guessing this is thanks to the screen size and/or aspect ratio difference. I bet they've got something that probably checks the aspect ratio on the screen and changes the duration based off of that value. So an iPhone 11 with a 2+ ratio compared to the ipad's 1.3 ratio will command a shorter duration. It looks like 0.25 is great for the ipad, but what happens if you did something where you made 0.25 * 1.333 (which is 0.33) the base and then use the device aspect ratio to determine the the actual duration. Placing the iPhone 11 closer to 0.15. Does that work?

**Comment by chinmaygarde** [comment]
I spent some more time on this yesterday and could not make this look consistent on iPad, iPhones and simulators. Moreover, the behavior looked slightly different on my iPhone X than on an old iPhone 6S. In all cases, there were no black bars or flashing but the subtle stretching was disconcerting still.

Since I am unlikely to spend more cycles on this in the short term, I am closing this PR.

**Pull request closed without merging**